pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentDispatcherSingleActiveConsumer.java

```diff
@@ -297,7 +297,11 @@ protected void readMoreEntries(Consumer consumer) {
                 log.debug("[{}-{}] Schedule read of {} messages", name, consumer, messagesToRead);
             }
             havePendingRead = true;
-            cursor.asyncReadEntriesOrWait(messagesToRead, this, consumer);
+            if (consumer.readCompacted()) {
+                topic.compactedTopic.asyncReadEntriesOrWait(cursor, messagesToRead, this, consumer);
```

Comment:
Once we finish reading the compacted snapshot, shouldn't the consumer switch to get messages from the point after the snapshot?

Reply:
it does. This is why we pass the cursor in. Otherwise we'd have something like
```java
if (consumer.readCompacted()) {
    if (topic.compactedTopic.shouldReadFromCompacted(cursor)) {
         topic.compactedTopic.asyncReadEntriesOrWait(cursor, messagesToRead, this, consumer);
    } else {
        // read from cursor as normal
    }
}
```
However, there could be another thread acting on compactedTopic, so the result from shouldReadFromCompacted may not be consistent with what happens when you try to read. i.e. a compaction could happen in between, and shouldReadFromCompacted could return false, even though in reality compacted data is available. This wouldn't be a huge issue, it just means more data is read.

Reply:
> it does. This is why we pass the cursor in. Otherwise we'd have something like

Got it, I had missed that part.

>  i.e. a compaction could happen in between

Shouldn't we be able to reset the subscription on the current snapshot and read from it. If there is another compaction ongoing, it should affect it. The data after the 1st snapshot would still be there. The only thing that might get complicated would be to delete the compacted ledger only after no one is using it anymore. This could a be an improvement for later anyway.

Reply:
well, when the other compaction finishes, it will send the acknowledgement to the broker, and this will update the ledger. We will read from this ledger from this point on.

I think this should also be the point where we delete the old ledger, as noone will be reading from it.